src/sqlancer/tidb/oracle/TiDBCERTOracle.java

```diff
@@ -0,0 +1,285 @@
+import sqlancer.tidb.ast.TiDBBinaryLogicalOperation;
+import sqlancer.tidb.ast.TiDBBinaryLogicalOperation.TiDBBinaryLogicalOperator;
+import sqlancer.tidb.ast.TiDBColumnReference;
+import sqlancer.tidb.ast.TiDBExpression;
+import sqlancer.tidb.ast.TiDBJoin;
+import sqlancer.tidb.ast.TiDBJoin.JoinType;
+import sqlancer.tidb.ast.TiDBSelect;
+import sqlancer.tidb.ast.TiDBTableReference;
+import sqlancer.tidb.visitor.TiDBVisitor;
+
+public class TiDBCERTOracle implements TestOracle<TiDBGlobalState> {
+    private final TiDBGlobalState state;
+    private TiDBExpressionGenerator gen;
+    private final ExpectedErrors errors = new ExpectedErrors();
+    private TiDBSelect select;
+    private List<String> queryPlan1Sequences;
+    private List<String> queryPlan2Sequences;
+
+    public TiDBCERTOracle(TiDBGlobalState globalState) {
+        state = globalState;
+        TiDBErrors.addExpressionErrors(errors);
+    }
+
+    @Override
+    public void check() throws SQLException {
+        queryPlan1Sequences = new ArrayList<>();
+        queryPlan2Sequences = new ArrayList<>();
+
+        // Randomly generate a query
+        TiDBTables tables = state.getSchema().getRandomTableNonEmptyTables();
+        gen = new TiDBExpressionGenerator(state).setColumns(tables.getColumns());
+        select = new TiDBSelect();
+
+        List<TiDBExpression> fetchColumns = new ArrayList<>();
+        fetchColumns.addAll(Randomly.nonEmptySubset(tables.getColumns()).stream().map(c -> new TiDBColumnReference(c))
+                .collect(Collectors.toList()));
+        select.setFetchColumns(fetchColumns);
+
+        List<TiDBExpression> tableList = tables.getTables().stream().map(t -> new TiDBTableReference(t))
+                .collect(Collectors.toList());
+        List<TiDBExpression> joins = TiDBJoin.getJoinsWithoutNature(tableList, state);
+        select.setJoinList(joins);
+        select.setFromList(tableList);
+        if (Randomly.getBoolean()) {
+            select.setWhereClause(gen.generateExpression());
+        }
+        if (Randomly.getBooleanWithRatherLowProbability()) {
+            select.setOrderByExpressions(gen.generateOrderBys());
+        }
+        if (Randomly.getBoolean()) {
+            select.setGroupByExpressions(select.getFetchColumns());
+            if (Randomly.getBoolean()) {
+                select.setHavingClause(gen.generateHavingClause());
+            }
+        }
+        if (Randomly.getBoolean()) {
+            select.setLimitClause(gen.generateExpression());
+        }
+        if (Randomly.getBoolean()) {
+            select.setOffsetClause(gen.generateExpression());
+        }
+
+        // Get the result of the first query
+        String queryString1 = TiDBVisitor.asString(select);
+        double rowCount1 = getRow(state, queryString1, queryPlan1Sequences);
+
+        // Mutate the query
+        boolean increase = false;
+        int mutator = (int) Randomly.getNotCachedInteger(0, 7);
+
+        switch (mutator) {
```

Comment:
We should use an enum for the different options here, which is less error prone.

Reply:
Solved